app.py

```diff
@@ -1,27 +1,114 @@
+from github_oauth import GithubOauthProvider
+from user.user import User
+from user.user_access import user_write_access, user_admin_access
+from user.user_role import UserRole
 
-def create_app(items = ItemStore()):
+def create_app(item_store = ItemStore(), user_store = UserStore()):
     app = Flask(__name__)
     logging.basicConfig(level=logging.INFO)
-    item_store = items
+
+    oauth_provider = GithubOauthProvider()
+
+    login_manager = LoginManager()
+    login_manager.init_app(app)
+
+    app.secret_key = oauth_provider.get_client_secret()
+    
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        # Authenticate with GitHub
+        return redirect(oauth_provider.get_authenticate_uri())
+    
+    @login_manager.user_loader
+    def load_user(user_id):
+        user = user_store.get_user(user_id)
+        return user
+
+    @app.route('/login/callback')
+    def login_callback():                                                            # pylint: disable=unused-variable
+        # Get authorization code from github 
+        code = request.args.get("code")
+
+        # Use the authorization code to get user info
+        userinfo = oauth_provider.get_user_info(
+            authorization_response=request.url,
+            redirect_url=request.base_url,
+            code=code)
+
+        user = User.from_json(userinfo)
+        user_store.add_user_if_missing(user)
+
+        # Begin user session
+        login_user(user)
+
+        return redirect("/")
+
+    @app.route("/logout")
+    @login_required
+    def logout():
+        logout_user()
+        return redirect("/")
+
+    @app.route("/admin")
+    @login_required
+    @user_admin_access
+    def admin():
+        users =  user_store.get_users()
+        return render_template('admin.html', users=users, login_disabled=is_login_disabled(),  current_user=current_user)
+
+    @app.route('/user/<operation>/<id>', methods=['POST'])
+    @login_required
+    @user_write_access
+    def update_user(operation, id):                                     # pylint: disable=unused-variable
+        print(f"Updateing user id={id} to {operation}")
+        if operation == "reader":
+            user_store.update_user(id, UserRole.READER)
+        elif operation == "writer":
+            user_store.update_user(id, UserRole.WRITER)
+        elif operation == "admin":
+            user_store.update_user(id, UserRole.ADMIN)
+        elif operation == "delete":
+            user_store.remove_user(id)
+        return redirect("/admin")
 
     @app.route('/')
+    @login_required
     def index():                                                            # pylint: disable=unused-variable
         items =  item_store.get_items()
-        return render_template('index.html', view_model=ViewModel(items))
-
+        return render_template('index.html', view_model=ViewModel(items), login_disabled=is_login_disabled(), current_user=current_user)
```

Comment:
Minor, but this `login_disabled` argument allows complexity to creep into the templates & make them harder to follow; this is a good example of where we could instead use a view-model for the current_user argument, which itself accounts for the possibility that login is disabled when checking if it's readonly (i.e. we pass a view model that wraps `current_user` and checks if the login is disabled as part of it's logic)